keps/sig-network/3015-prefer-local-traffic-policy/README.md

```diff
@@ -0,0 +1,548 @@
+
+Since someone using `externalTrafficPolicy: PreferLocal` cannot _rely_
+on always having the correct source IP, it would probably be least
+confusing to just say that `PreferLocal` services _never_ preserve the
+source IP, just like `externalTrafficPolicy: Cluster` services don't.
+Thus, the behavior would be:
+
+- `Cluster`: Prioritizes load-spreading. A node receiving a connection
+  to a `Cluster` service is equally likely to send it to any of the
+  service's endpoints (with no preference given to local endpoints).
+  Use this when the clients or intermediaries do not balance
+  connections between nodes themselves. (eg, perhaps you are using a
+  `NodePort` service, but only advertising a single node IP for
+  clients to connect to). Does not preserve source IP.
+
+- `Local`: Prioritizes efficiency and preservation of the source IP. A
+  node receiving a connection to a `Local` service will send it to one
+  of the service's endpoints on the same node; if there are no
+  endpoints on the same node then the packets will be discarded.
+  (Hopefully the client will try again and hit a valid node the next
+  time.) Since packets never leave the node, the original source IP
+  can be preserved. Depends on an upstream load balancer to distribute
+  connections evenly between nodes (and to not send connections to
+  nodes with no endpoints).
+
+- `PreferLocal`: Prioritizes efficiency and reliability by combining
+  aspects of the other two policies. A node receiving a connection to
+  a `PreferLocal` service will send it to one of the service's
+  endpoints on the same node, if possible; if there are no local
+  endpoints then it will send it to a random remote endpoint, as with
+  `Cluster`. Like `Local` it depends on an upstream load balancer to
+  distribute connections evenly, but it is slightly more reliable in
+  the face of changing endpoints, at the expense of not preserving the
+  source IP (even when delivering locally).
+
+```
+<<[UNRESOLVED renaming ]>>
+
+Should we add a new alias for `Local`, such as `PreserveSourceIP`, and
+deprecate the old name?
+
+<<[/UNRESOLVED]>>
+```
+
+[public documentation on externalTrafficPolicy]: https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip
+
+### Internal Traffic Policy
+
+This is easy; source IP is always preserved for internal traffic, so
+`PreferLocal` can be implemented in the obvious way with no caveats.
+
+```
+<<[UNRESOLVED internal-traffic-policy-local ]>>
+
+Given `PreferLocal`, is there any reason to keep the behavior of
+`internalTrafficPolicy: Local` with no fallback? Is there actually a
+use case for "either send locally or drop the traffic"?
+
+The KEP gives the use case:
+
+    As a platform owner, I want to create a Service that always
+    directs traffic to a logging daemon on the same node. Traffic
+    should never bounce to a daemon on another node.
```

Comment:
Singleton pattern? I honestly tried to find an example and I couldn't

Reply:
I'm not sure what you're saying.

Reply:
this can be useful if you want to have a service with special characteristics per node... I can't find an example neither, I was thinking if maybe someone want to have an unique service in the cluster that only works in one node ...